crates/hir/src/has_source.rs

```diff
@@ -136,8 +136,8 @@ impl HasSource for Impl {
     }
 }
 
-impl HasSource for TypeParam {
-    type Ast = Either<ast::TypeParam, ast::Trait>;
+impl HasSource for TypeOrConstParam {
+    type Ast = Either<Either<ast::TypeParam, ast::ConstParam>, ast::Trait>;
```

Comment:
What a lovely type, I think having a `ast::TypeOrConstParam` enum wrapper in `syntax` would be a bit nicer to have